gateway/src/resty/synchronization.lua

```diff
@@ -53,4 +53,28 @@ function _M:release(key)
   semaphores[key] = nil
 end
 
+--- run a new function, callback using locks on the given key
+-- @tparam string key: key for the semaphore
+-- @tparam int timeout: timeout for getting the lock before raise an error.
+-- @tparam function callback: function to execute if the lock is adquired correctly
+function _M:run(key, timeout, callback, ...)
+  local sema = self:acquire(key)
+
+  local ok, err = sema:wait(timeout)
+  if not ok then
+    ngx.log(ngx.WARN, 'failed to acquire lock on key: ', key, ' error: ', err)
+    return false, nil
+  end
+
+  local task = safe_task.new(callback)
+  local ret, result = task:execute(...)
+
+  if ok then
+    self.release(key)
+    sema:post()
+  end
+
+  return ret, result
```

Comment:
Are 2 enough? According to the docs of the function, `execute` returns 2 or 3 params depending on whether there was an error or not.

Reply:
When there's an error, `execute` returns 3 params and the error msg is the third one, so I think we'll lose it.
https://github.com/3scale/APIcast/blob/6b42baa85c9bde860c4c3bf70e8af78b48a710cf/gateway/src/resty/concurrent/safe_task_executor.lua#L39